tap_github/client.py

```diff
@@ -266,10 +271,86 @@ def parse_response(self, response: requests.Response) -> Iterable[dict]:
         resp_json = response.json()
         yield from extract_jsonpath(self.query_jsonpath, input=resp_json)
 
+    def get_next_page_token(
+        self, response: requests.Response, previous_token: Optional[Any]
+    ) -> Optional[Any]:
+        """
+        Return a dict of cursors for identifying next page or None if no more pages.
+
+        Note - pagination requires the Graphql query to have nextPageCursor_X parameters
+        with the assosciated hasNextPage_X, startCursor_X and endCursor_X.
+
+        X should be an integer between 0 and 9, increasing with query depth.
+        """
+
+        resp_json = response.json()
+        # Find if results contains "hasNextPage_X" flags and if any are True.
+        # If so, set nextPageCursor_X to endCursor_X for X max.
+
+        next_page_results = collections.OrderedDict(
+            nested_lookup(
+                key="hasNextPage",
+                document=resp_json,
+                wild=True,
+                with_keys=True,
+            )
+        )
+
+        has_next_page_results = collections.OrderedDict()
+        # Iterate over all the items and filter items with hasNextPage = True.
+        for (key, value) in next_page_results.items():
+            # Check if key is even then add pair to new dictionary
+            if any(value):
+                has_next_page_results[key] = value
+
+        # Check if any "hasNextPage" is True.
+        resp_has_next_page = bool(has_next_page_results)
+
+        # No pagination necessary, exit early.
+        if not resp_has_next_page:
+            return None
+
+        # Get deepest pagination item
+        (pagination_key, _) = next(reversed(has_next_page_results.items()))
+        pagination_key_number = re.findall(r"\d+", pagination_key)[0]
+
+        next_page_start_cursor_results = nested_lookup(
+            key=f"startCursor",
+            document=resp_json,
+            wild=True,
+            with_keys=True,
+        )
+
+        next_page_end_cursor_results = nested_lookup(
+            key=f"endCursor_{pagination_key_number}"
+            if pagination_key_number
+            else "endCursor",
+            document=resp_json,
+            with_keys=True,
+        )
+
+        # Set all startCursor to their previous value.
+        # Then only increment the deepest one with endCursor
+        next_page_cursors = dict()
+        for (key, value) in list(next_page_start_cursor_results.items()) + list(
+            next_page_end_cursor_results.items()
+        ):
+            start_pagination_key_number = re.findall(r"\d+", key)[0]
```

Comment:
Since this regexp is going to run quite a few times, you could `re.compile()` it on class instantiation, and then reuse that. Not a lifechanger, but might as well.
Or since the string seem to always look like `something_N` you could also use `str.split('_')` which might be faster?